core/amber/src/main/scala/edu/uci/ics/texera/web/resource/dashboard/user/dataset/service/GitVersionControlLocalFileStorageService.java

```diff
@@ -0,0 +1,238 @@
+   * Writes the file content to the provided OutputStream.
+   *
+   * @param baseRepoPath The repository path.
+   * @param commitHash The commit hash of the version from which the file content is retrieved.
+   * @param filePath The path of the file within the repository.
+   * @param outputStream The OutputStream to which the file content is written.
+   * @throws IOException If an I/O error occurs.
+   * @throws InterruptedException If the operation is interrupted.
+   */
+  public static void retrieveFileContentOfVersion(String baseRepoPath, String commitHash, String filePath, OutputStream outputStream) throws IOException, InterruptedException {
+    GitSystemCall.showFileContentOfCommit(baseRepoPath, commitHash, filePath, outputStream);
+  }
+
+  /**
+   * Recovers the repository to its latest version, discarding any uncommitted changes.
+   *
+   * @param baseRepoPath The repository path.
+   * @throws IOException If an I/O error occurs.
+   * @throws InterruptedException If the operation is interrupted.
+   */
+  public static void recoverToLatestVersion(String baseRepoPath) throws IOException, InterruptedException {
+    if (GitSystemCall.hasUncommittedChanges(baseRepoPath)) {
+      GitSystemCall.rollbackToLatestCommit(baseRepoPath);
+    }
+  }
+
+  /**
+   * Utility class for executing Git commands in a system-agnostic manner.
+   */
+  static class GitSystemCall {
+
+    public static Charset BYTE_2_STRING_CHARSET = StandardCharsets.UTF_8;
+
+    // utility function to execute git-related system call
+    private static byte[] executeGitCommand(String workingDirectory, String... args) throws IOException, InterruptedException {
+      List<String> commands = new ArrayList<>();
+      commands.add("git");
+      Collections.addAll(commands, args);
+
+      ProcessBuilder builder = new ProcessBuilder(commands);
+      builder.directory(new File(workingDirectory));
+      builder.redirectErrorStream(true);
+      Process process = builder.start();
+
+      try (ByteArrayOutputStream outputStream = new ByteArrayOutputStream()) {
+        byte[] buffer = new byte[1024];
+        int length;
+        while ((length = process.getInputStream().read(buffer)) != -1) {
+          outputStream.write(buffer, 0, length);
+        }
+
+        int exitCode = process.waitFor();
+        if (exitCode != 0) {
+          throw new IOException("Failed to execute Git command: " + String.join(" ", commands));
+        }
+
+        return outputStream.toByteArray();
+      }
+    }
+
+    public static void initRepo(String path) throws IOException, InterruptedException {
+      executeGitCommand(path, "init");
+    }
+
+    // this function adds all changes to the stage and create a commit, it will return the hash of that commit
+    public static String addAndCommit(String repoPath, String commitMessage) throws IOException, InterruptedException {
+      // Adding all files and committing
+      executeGitCommand(repoPath, "add", ".");
```

Comment:
I do not think it is a good idea to do `add .` as it will involve all the files in the current directory. It's better to let the user explicitly add files than add a directory, which may include unnecessary files.